crates/adapters/src/controller/stats.rs

```diff
@@ -1310,6 +1410,233 @@ impl TokenList {
+    total_processed_records: u64,
+
+    /// Not-yet-completed watermarks, in the order they were added.
+    watermark_list: VecDeque<WatermarkListEntry>,
+
+    /// The latest completed watermark with non-empty metadata, if any.
+    completed_watermark: Option<CompletedWatermark>,
+}
+
+impl WatermarkTrackerInner {
+    fn new() -> Self {
+        Self {
+            total_completed_records: 0,
+            total_processed_records: 0,
+            watermark_list: VecDeque::new(),
+            completed_watermark: None,
+        }
+    }
+
+    fn append(&mut self, watermarks: Vec<Watermark>, global_offset: u64) {
+        for watermark in watermarks {
+            // Keep the queue bounded. This gives older watermarks
+            // a chance to complete; otherwise we risk always evicting all watermarks before the complete
+            // and never reporting any completed watermarks.
+            if self.watermark_list.len() >= MAX_TRACKED_WATERMARKS {
+                break;
+            }
+            self.watermark_list.push_back(WatermarkListEntry {
+                watermark,
+                global_offset,
+                processed_at: None,
+            });
+        }
+    }
+
+    fn update_processed(&mut self, processed_records: u64, ts: DateTime<Utc>) {
+        self.total_processed_records = processed_records;
+
+        for entry in self.watermark_list.iter_mut() {
+            if entry.global_offset <= processed_records && entry.processed_at.is_none() {
+                entry.processed_at = Some(ts);
+            } else {
+                break;
+            }
+        }
+    }
+
+    fn update_completed(
+        &mut self,
+        completed_records: u64,
+        ts: DateTime<Utc>,
+        metrics: &InputEndpointMetrics,
+    ) {
+        if self.total_completed_records >= completed_records {
+            return;
+        }
+        self.total_completed_records = completed_records;
+
+        while let Some(entry) = self.watermark_list.front() {
+            if entry.global_offset <= completed_records {
+                let entry = self.watermark_list.pop_front().unwrap();
+
+                let processed_at = entry.processed_at.unwrap_or(ts);
+
+                let _ = metrics
+                    .processing_latency_millis_histogram
+                    .lock()
+                    .map(|mut histogram| {
+                        histogram.record(
+                            processed_at.timestamp_millis()
+                                - entry.watermark.timestamp.timestamp_millis(),
+                        )
+                    });
+                let _ = metrics
+                    .completion_latency_millis_histogram
+                    .lock()
+                    .map(|mut histogram| {
+                        histogram.record(
+                            ts.timestamp_millis() - entry.watermark.timestamp.timestamp_millis(),
+                        )
+                    });
```

Comment:
These would be more clearly expressed as `if let Ok(histogram) = ... { histogram.record(...); }`.